StandardFiniteFieldsOSCAR.jl

```diff
@@ -1,443 +1,443 @@
+ function Base.setproperty!(F::FinField, v::Symbol, c)
+     if v == :primitivepowersintowerbasis
+         if !isdefined(F, :__attrs)
+             setfield!(F, :__attrs, Dict{Symbol, Any}(v => c))
+             push!(getfield(F, :__attrs), :towerbasis => inv(c))
+         else
+             push!(getfield(F, :__attrs), v =>c)
+             push!(getfield(F, :__attrs), :towerbasis => inv(c))
+         end
+     elseif v == :towerbasis
+         error()
+     elseif v in _propertysymbols
+         if !isdefined(F, :__attrs)
+             setfield!(F, :__attrs, Dict{Symbol, Any}(v => c))
+         else
+             push!(getfield(F, :__attrs), v =>c)
+         end
+     else
+         setfield!(F, v, c)
+     end
+ end
+
+ function Base.getproperty(F::FinField, v::Symbol)
+     if v == :is_standardfinitefield
+         if !isdefined(F, :__attrs) || !haskey(getfield(F, :__attrs), :is_standardfinitefield)
+             return false
+         else
+             return get(getfield(F, :__attrs), :is_standardfinitefield, false)
+         end
+     elseif v == :is_standardprimefield
+         if !isdefined(F, :__attrs) || !haskey(getfield(F, :__attrs), :is_standardprimefield)
+             return false
+         else
+             return get(getfield(F, :__attrs), :is_standardprimefield, false)
+         end
+     elseif v in _propertysymbols
+         if !isdefined(F, :__attrs)
+             error()
+         else
+             return get(getfield(F, :__attrs), v, nothing)
+         end
+     else
+         return getfield(F, v)
+     end
+ end
+
+
+ # TODO: Lübeck speeds this up by caching triples [q,m,a] resulting from this
+ function standardaffineshift(q::ZZRingElem, i)
+     m = div(4*q, 5)
+     while gcd(m,q) != 1
+         m -= 1
+     end
+     a = div(2*q, 3)
+     return mod((m*i + a), q)
+ end
+
+ # Given a field F and Steinitz number n, give the corresponding field element.
+ # we REQUIRE that F is a standard finite field TODO: using @assert?
+ function elementfromsteinitznumber(F::FinField, n)
+     p = characteristic(F)
+     q = order(F)
+     if n < 0 || n > q
+         error("We need to have 0 <= n <= q")
+     end
+     if n == 0
+         return zero(F)
+     elseif q == p
+         return F(n)
+     else
+         vectorrep = GF(p).(digits(n, base = Int(p)))
+         # this forms a linear combo of F.towervasis rows using vectorrep as coefficients,
+         # and then convert this vector to an element of F.
+         return F(vectorrep * @view F.towerbasis[1:length(vectorrep), :])
+     end
+ end
+
+ # Returns an element a in F that is NOT an rth root of unity
+ # we REQUIRE that F is a standard finite field TODO: using @assert?
+ function non_rth_root(F::FinField, r)
+     q = order(F)
+     if mod(q-1, r) == 0
+         i = 0
+         a = zero(F)
+         while a == 0 || a^(div(q-1,r)) == one(F)
```

Comment:
Pull out the constant, maybe?
```suggestion
         k = div(q-1,r)
         while iszero(a) || isone(a^k)
```